src/query/expression/src/evaluator.rs

```diff
@@ -502,6 +505,85 @@ impl<'a> Evaluator<'a> {
                 }
                 other => unreachable!("source: {}", other),
             },
+            (DataType::Variant, DataType::Array(inner_dest_ty)) => {
+                let empty_vec = vec![jsonb::Value::Null];
+                match value {
+                    Value::Scalar(Scalar::Variant(x)) => {
+                        let array = jsonb::from_slice(&x).map_err(|e| {
+                            ErrorCode::BadArguments(format!(
+                                "Expect to be valid json, got err: {e:?}"
+                            ))
+                        })?;
+                        let array = array.as_array().unwrap_or(&empty_vec);
```

Comment:
if the value is not a json array, we'd better cast it as an empty array instead of an array with json null, this may lead to unexpected results.
for example:
```sql
MySQL [(none)]> select json_array_transform(parse_json('"aa"'), data -> CONCAT(data::String, 'bend'));
+--------------------------------------------------------------------------------+
| json_array_transform(parse_json('"aa"'), data -> concat(data::STRING, 'bend')) |
+--------------------------------------------------------------------------------+
| ["nullbend"]                                                                   |
+--------------------------------------------------------------------------------+
1 row in set (0.048 sec)
```